biggraphite/cli/command_copy.py

```diff
@@ -0,0 +1,149 @@
+
+
+class CommandCopy(command.BaseCommand):
+    """Copy for metrics."""
+
+    NAME = "copy"
+    HELP = "Copy metrics."
+
+    def add_arguments(self, parser):
+        """Add custom arguments.
+
+        See command.CommandBase.
+        """
+        parser.add_argument(
+            "src",
+            help="One source metric or subdirectory name"
+        )
+        parser.add_argument(
+            "dst",
+            help="One destination metric or subdirectory name"
+        )
+        parser.add_argument(
+            "-r",
+            "--recursive",
+            action="store_const", default=False, const=True,
+            help="Copy points for all metrics as a subtree",
+        )
+        parser.add_argument(
+            "--time-start",
+            action=command.ParseDateTimeArg,
+            help="Copy points written later than this time.",
+            default=datetime.datetime.now() - datetime.timedelta(minutes=10),
+            required=False,
+        )
+        parser.add_argument(
+            "--time-end",
+            action=command.ParseDateTimeArg,
+            help="Copy points written earlier than this time.",
+            default=datetime.datetime.now(),
+            required=False,
+        )
+        parser.add_argument(
+            "--dry-run",
+            action="store_const", default=False, const=True,
+            help="Only show commands to create/upgrade the schema.",
+            required=False,
+        )
+        parser.add_argument(
+            "--log-level",
+            default="INFO",
+            help="Override log levels.",
+            required=False,
+        )
+        parser.add_argument(
+            "--max-metrics-per-pattern",
+            metavar="MAX", type=int,
+            help="Override accessor max_metrics_per_pattern attribute.",
+            required=False,
+        )
+
+    def run(self, accessor, opts):
+        """Copy metrics and directories.
+
+        See command.CommandBase.
+        """
+        accessor.connect()
+
+        log.setLevel(opts.log_level)
+
+        if accessor.max_metrics_per_pattern and opts.max_metrics_per_pattern:
+            accessor.max_metrics_per_pattern = opts.max_metrics_per_pattern
+
+        metric_tuples = self._get_metric_tuples(
+            accessor, opts.src, opts.dst, opts.recursive, opts.dry_run)
+        if opts.dry_run:
+            for m, _ in metric_tuples:
+                log.info("Metric to copy : %s" % m.name)
+            return
+
+        time_start = time.mktime(opts.time_start.timetuple())
+        time_end = time.mktime(opts.time_end.timetuple())
+
+        for src_metric, dst_metric in metric_tuples:
+            self._copy_metric(accessor, src_metric, dst_metric, time_start, time_end)
+
+    @staticmethod
+    def _get_metric_tuples(accessor, src, dst, recursive, dry_run=True):
+        pattern = "%s.**" % src if recursive else src
+        try:
+            src_metrics = list_metrics(accessor, pattern)
```

Comment:
list_metrics() returns a list, not a generator. If you want to make it work with large numbers of metrics you'll need to change it to return a generator